pkg/agent/flowexporter/connections/conntrack_ovs.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connections
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/config"
+	"github.com/vmware-tanzu/antrea/pkg/agent/flowexporter"
+	"github.com/vmware-tanzu/antrea/pkg/ovs/ovsctl"
+	"github.com/vmware-tanzu/antrea/pkg/util/ip"
+)
+
+// connTrackOvsCtl implements ConnTrackDumper. This supports OVS userspace datapath scenarios.
+var _ ConnTrackDumper = new(connTrackOvsCtl)
+
+type connTrackOvsCtl struct {
+	nodeConfig   *config.NodeConfig
+	serviceCIDR  *net.IPNet
+	ovsctlClient ovsctl.OVSCtlClient
+}
+
+func NewConnTrackOvsAppCtl(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet, ovsctlClient ovsctl.OVSCtlClient) *connTrackOvsCtl {
+	if ovsctlClient == nil {
+		return nil
+	}
+	return &connTrackOvsCtl{
+		nodeConfig,
+		serviceCIDR,
+		ovsctlClient,
+	}
+}
+
+// DumpFlows uses "ovs-appctl dpctl/dump-conntrack" to dump conntrack flows in the Antrea ZoneID.
+func (ct *connTrackOvsCtl) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connection, error) {
+	conns, err := ct.ovsAppctlDumpConnections(zoneFilter)
+	if err != nil {
+		klog.Errorf("Error when dumping flows from conntrack: %v", err)
+		return nil, err
+	}
+
+	filteredConns := filterAntreaConns(conns, ct.nodeConfig, ct.serviceCIDR, zoneFilter)
+	klog.V(2).Infof("Flow exporter considered flows: %d", len(filteredConns))
+
+	return filteredConns, nil
+}
+
+func (ct *connTrackOvsCtl) ovsAppctlDumpConnections(zoneFilter uint16) ([]*flowexporter.Connection, error) {
+	// Dump conntrack using ovs-appctl dpctl/dump-conntrack
+	cmdOutput, execErr := ct.ovsctlClient.RunAppctlCmd("dpctl/dump-conntrack", false, "-m", "-s")
+	if execErr != nil {
+		return nil, fmt.Errorf("error when executing dump-conntrack command: %v", execErr)
+	}
+
+	// Parse the output to get the flows
+	antreaConns := make([]*flowexporter.Connection, 0)
+	outputFlow := strings.Split(string(cmdOutput), "\n")
+	var err error
+	for _, flow := range outputFlow {
```

Comment:
putting a sample flow here as a comment would help reading this:
```
# tcp,orig=(src=10.10.1.2,dst=10.96.0.1,sport=42540,dport=443),reply=(src=10.96.0.1,dst=10.10.1.2,sport=443,dport=42540),zone=65520,protoinfo=(state=TIME_WAIT)
```

BTW, is the order of the fields not guaranteed here? If it is, I think this code is overly complicated, you could have this:
```
flowSlice := strings.Split(flow, ",")
proto := flowSlice[0]
# ...
orig := strings.Split(flowSlice[1], "=")
# ...
```
or a regex match would do as well

for this type of code, rather than use `klog.Errorf(...); continue` for every possible error, I think you should define a helper function to convert a flow string to a `flowexporter.Connection` connection object and return in case of error. Then in this function you can log the error if there is one.

Reply:
Sometimes some fields will be missing in the flow string. Example: Not all fields will have zone; there could be other scenarios too when some fields are missing.
This definitely needs a revisit. We do not add stats as well. Currently, it is used for e2e test on kind clusters. I will try to simplify this in a follow up patch. Hope that's ok.
Handled the error scenario as suggested.